compiler/optimizations/deadCodeElimination.cpp

```diff
@@ -905,6 +916,99 @@ static void cleanupLoopBlocks(FnSymbol* fn) {
   }
 }
 
+extern int classifyPrimitive(CallExpr *call, bool inLocal);
+extern bool inLocalBlock(CallExpr *call);
+
+bool debugPrint = false; 
+bool allowFnCallsFromGPU = false;
+int indent = 0;
+
+static bool blockLooksLikeStreamForGPU(BlockStmt* blk, bool allowFnCalls) {
+  if (!blk->inTree() || blk->getModule()->modTag != MOD_USER)
+    return false;
+
+  if (CForLoop* cfl = toCForLoop(blk))
+    if (!cfl->isOrderIndependent())
+      return false;
+
+  std::set<FnSymbol*> okFns;
+  std::set<FnSymbol*> visitedFns;
+
+  return blockLooksLikeStreamForGPUHelp(blk, okFns, visitedFns, allowFnCalls);
+}
+
+static bool blockLooksLikeStreamForGPUHelp(BlockStmt* blk, std::set<FnSymbol*>& okFns, std::set<FnSymbol*> visitedFns, bool allowFnCalls) {
+
+  if (debugPrint) {
+    FnSymbol* fn = blk->getFunction();
+    printf("%*s%s:%d: %s[%d]\n", indent, "", fn->fname(), fn->linenum(), fn->name, fn->id);
+  }
+
+  if (visitedFns.count(blk->getFunction()) != 0) {
+    if (blk->getFunction()->hasFlag(FLAG_FUNCTION_TERMINATES_PROGRAM)) {
+      return true; // allow `halt` to be potentially called recursively
+    } else {
+      return false;
+    }
+  }
+
+  visitedFns.insert(blk->getFunction());
+
+  std::vector<CallExpr*> calls;
+  collectCallExprs(blk, calls);
+
+  for_vector(CallExpr, call, calls) {
+    if (call->primitive) {
+      // only primitives that are fast and local are allowed for now
+      bool inLocal = inLocalBlock(call);
+      int is = classifyPrimitive(call, inLocal);
+      if ((is != FAST_AND_LOCAL)) {
```

Comment:
I think relying on `FAST_AND_LOCAL` is good enough for now. But I think we'll have to do our own classification for GPU. I am not sure if there's a strict relation between the ability of running a primitive on an AM handler (`FAST`) vs running it on GPU.